File: app/src/main/java/org/ametro/ui/widgets/MapBottomPanelWidget.java
package org.ametro.ui.widgets;

import android.animation.Animator;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import org.ametro.R;
import org.ametro.app.Constants;
import org.ametro.model.entities.MapSchemeLine;
import org.ametro.model.entities.MapSchemeStation;

public class MapBottomPanelWidget implements Animator.AnimatorListener {

    private final View view;

    private final TextView stationTextView;
    private final TextView lineTextView;

    private final Button detailButton;
    private final Button beginButton;
    private final Button endButton;

    private final IMapBottomPanelEventListener listener;

    private final Runnable hideAnimation = new Runnable() {
        @Override
        public void run() {
            view.animate().setDuration(Constants.ANIMATION_DURATION).setListener(MapBottomPanelWidget.this).translationY(view.getHeight());
        }
    };

    private final Runnable showAnimation = new Runnable() {
        @Override
        public void run() {
            view.setVisibility(View.VISIBLE);
            stationTextView.setText(station.getDisplayName());
            lineTextView.setText(line.getDisplayName());
            view.animate().setDuration(Constants.ANIMATION_DURATION).setListener(MapBottomPanelWidget.this).translationY(0);
        }
    };

    private Runnable actionOnEndAnimation;
    private boolean visible;
    private boolean firstTime;

    private MapSchemeLine line;
    private MapSchemeStation station;

    public MapBottomPanelWidget(ViewGroup view, IMapBottomPanelEventListener listener) {
        this.view = view;
        this.listener = listener;
        this.visible = false;
        this.firstTime = true;

        stationTextView = (TextView) view.findViewById(R.id.station);
        lineTextView = (TextView) view.findViewById(R.id.line);
        detailButton = (Button) view.findViewById(R.id.button_details);
        beginButton = (Button) view.findViewById(R.id.button_begin);
        endButton = (Button) view.findViewById(R.id.button_end);

        View.OnClickListener clickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                if (v == detailButton) {
                    MapBottomPanelWidget.this.listener.onShowMapDetail(line, station);
                } else if (v == beginButton) {
                    MapBottomPanelWidget.this.listener.onSelectBeginStation(line, station);
                } else if (v == endButton) {
                    MapBottomPanelWidget.this.listener.onSelectEndStation(line, station);
                }
            }
        };

        view.setOnClickListener(clickListener);
        detailButton.setOnClickListener(clickListener);
        beginButton.setOnClickListener(clickListener);
        endButton.setOnClickListener(clickListener);
    }

    public boolean isOpened() {
        return visible;
    }

    public void show(final MapSchemeLine line, final MapSchemeStation station, boolean showDetails) {

        detailButton.setVisibility(showDetails ? View.VISIBLE : View.INVISIBLE);

        if (visible && this.line == line && this.station == station) {
            return;
        }

        this.line = line;
        this.station = station;

        if (!visible && !firstTime) {
            visible = true;
            showAnimation.run();
            return;
        }

        visible = true;
        firstTime = false;
        actionOnEndAnimation = showAnimation;
        hideAnimation.run();
    }

    public void hide() {
        if (!visible) {
            return;
        }
        visible = false;
        hideAnimation.run();
    }

    @Override
    public void onAnimationStart(Animator animation) {
    }

    @Override
    public void onAnimationEnd(Animator animation) {
        if (!visible)
            view.setVisibility(View.INVISIBLE);

        if (actionOnEndAnimation != null) {
            actionOnEndAnimation.run();
            actionOnEndAnimation = null;
        }
    }

    @Override
    public void onAnimationCancel(Animator animation) {
    }

    @Override
    public void onAnimationRepeat(Animator animation) {

    }

    public interface IMapBottomPanelEventListener {
        void onShowMapDetail(MapSchemeLine line, MapSchemeStation station);

        void onSelectBeginStation(MapSchemeLine line, MapSchemeStation station);

        void onSelectEndStation(MapSchemeLine line, MapSchemeStation station);
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/widgets/MapBottomPanelWidget.kt
package io.github.romangolovanov.apps.ametro.ui.widgets

import android.animation.Animator
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import io.github.romangolovanov.apps.ametro.R
import io.github.romangolovanov.apps.ametro.app.Constants
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeLine
import io.github.romangolovanov.apps.ametro.model.entities.MapSchemeStation

class MapBottomPanelWidget(viewGroup: ViewGroup, private val listener: IMapBottomPanelEventListener) :
    Animator.AnimatorListener {

    private val view: View = viewGroup

    private val stationTextView: TextView = viewGroup.findViewById(R.id.station)
    private val lineTextView: TextView = viewGroup.findViewById(R.id.line)

    private val detailButton: Button = viewGroup.findViewById(R.id.button_details)
    private val beginButton: Button = viewGroup.findViewById(R.id.button_begin)
    private val endButton: Button = viewGroup.findViewById(R.id.button_end)

    private val hideAnimation = Runnable {
        view.animate()
            .setDuration(Constants.ANIMATION_DURATION)
            .setListener(this@MapBottomPanelWidget)
            .translationY(view.height.toFloat())
    }

    private val showAnimation = Runnable {
        view.visibility = View.VISIBLE
        stationTextView.text = station!!.displayName
        lineTextView.text = line!!.displayName
        view.animate()
            .setDuration(Constants.ANIMATION_DURATION)
            .setListener(this@MapBottomPanelWidget)
            .translationY(0f)
    }

    private var actionOnEndAnimation: Runnable? = null
    private var visible = false
    private var firstTime = true

    private var line: MapSchemeLine? = null
    private var station: MapSchemeStation? = null

    init {
        val clickListener = View.OnClickListener { v ->
            when (v) {
                detailButton -> listener.onShowMapDetail(line!!, station!!)
                beginButton -> listener.onSelectBeginStation(line!!, station!!)
                endButton -> listener.onSelectEndStation(line!!, station!!)
            }
        }

        viewGroup.setOnClickListener(clickListener)
        detailButton.setOnClickListener(clickListener)
        beginButton.setOnClickListener(clickListener)
        endButton.setOnClickListener(clickListener)
    }

    val isOpened: Boolean get() = visible

    fun show(line: MapSchemeLine, station: MapSchemeStation, showDetails: Boolean) {
        detailButton.visibility = if (showDetails) View.VISIBLE else View.INVISIBLE

        if (visible && this.line == line && this.station == station) return

        this.line = line
        this.station = station

        if (!visible && !firstTime) {
            visible = true
            showAnimation.run()
            return
        }

        visible = true
        firstTime = false
        actionOnEndAnimation = showAnimation
        hideAnimation.run()
    }

    fun hide() {
        if (!visible) return
        visible = false
        hideAnimation.run()
    }

    override fun onAnimationStart(animation: Animator) {}

    override fun onAnimationEnd(animation: Animator) {
        if (!visible) view.visibility = View.INVISIBLE
        actionOnEndAnimation?.run()
        actionOnEndAnimation = null
    }

    override fun onAnimationCancel(animation: Animator) {}

    override fun onAnimationRepeat(animation: Animator) {}

    interface IMapBottomPanelEventListener {
        fun onShowMapDetail(line: MapSchemeLine, station: MapSchemeStation)
        fun onSelectBeginStation(line: MapSchemeLine, station: MapSchemeStation)
        fun onSelectEndStation(line: MapSchemeLine, station: MapSchemeStation)
    }
}
